### PR TITLE
xcm crate

### PR DESCRIPTION
**Pull request by justinzhou93**

### CODE REVIEW

File: pallets/template/src/xcm_test.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
```
Comment (rhuttman) [inline pallets/template/src/xcm_test.rs:3]: is this file needed?

File: pallets/template/src/lib.rs [hunk 2]
```diff
@@ -14,92 +22,155 @@ mod tests;
...
 		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type Origin: From<<Self as SystemConfig>::Origin>
+			+ Into<Result<CumulusOrigin, <Self as Config>::Origin>>;
+
+		type Call: From<Call<Self>> + Encode;
+
+		type XcmSender: SendXcm;
+		type XcmExecutor: ExecuteXcm<<Self as pallet::Config>::Call>;
+		// type WeightInfo: pallet_automation_time::WeightInfo;
```
Comment (rhuttman) [inline pallets/template/src/lib.rs:51]: remove if not needed

File: pallets/template/src/lib.rs [hunk 3]
```diff
@@ -14,92 +22,155 @@ mod tests;
...
-			let who = ensure_signed(origin)?;
-
-			// Update storage.
-			<Something<T>>::put(something);
-
-			// Emit an event.
-			Self::deposit_event(Event::SomethingStored(something, who));
-			// Return a successful DispatchResultWithPostInfo
-			Ok(().into())
+		pub fn send_remark_with_event(origin: OriginFor<T>, para: ParaId, message: Vec<u8>) -> DispatchResult {
```
Comment (rhuttman) [inline pallets/template/src/lib.rs:86]: this still needed?

File: pallets/template/src/lib.rs [hunk 4]
```diff
@@ -14,92 +22,155 @@ mod tests;
...
+		/**
+		 * This function implements XCM call to OAK with the OAK XCM crate. It uses the `force_send_balance` extrinsic above.
+		 * We can create an XCMP instruction with that call wrapped in the instructions to be sent back to this chain.
+		 * This implementation withdraws assets for the fee from the sovereign account of this chain on Turing.
+		 * Therefore, TUR tokens must be available for this sovereign account on the Turing chain. 
+		 */
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+		pub fn send_schedule_xcmp_with_crate(
+			origin: OriginFor<T>,
+			para: ParaId,
```
Comment (rhuttman) [inline pallets/template/src/lib.rs:140]: This doesn;t seem to be needed as the crate knows where to send it

File: pallets/template/src/lib.rs [hunk 5]
```diff
@@ -14,92 +22,155 @@ mod tests;
...
+			para_response_location: ParaId,
```
Comment (rhuttman) [inline pallets/template/src/lib.rs:141]: why is this needed? currently you can only send back to your own chain.

File: pallets/template/src/lib.rs [hunk 6]
```diff
@@ -14,92 +22,155 @@ mod tests;
...
+			provided_id: Vec<u8>,
+			execution_times: Vec<u64>,
+			source: T::AccountId,
```
Comment (rhuttman) [inline pallets/template/src/lib.rs:144]: the source would be the caller. Make sense to remove this as a sudo call and just use the caller's accountId. Else your force name above would be valid

File: pallets/template/src/lib.rs [hunk 7]
```diff
@@ -14,92 +22,155 @@ mod tests;
...
+			ensure_root(origin)?;
+			let call_name = b"automation_time_schedule_xcmp_with_crate".to_vec();
+			let inner_call = <T as Config>::Call::from(Call::<T>::force_send_balance { source: source.clone(), dest, value })
+				.encode()
+				.into();
+			let transact_instruction =
+				T::OakXcmInstructionGenerator::create_schedule_xcmp_instruction(provided_id, execution_times, para_response_location, inner_call);
+			let asset = MultiAsset {
+				id: Concrete(MultiLocation::here()),
+				fun: Fungibility::Fungible(7_000_000_000),
```
Comment (rhuttman) [inline pallets/template/src/lib.rs:157]: this is hard coded until you add the fee stuff to the crate right?

File: pallets/template/src/lib.rs [hunk 8]
```diff
@@ -14,92 +22,155 @@ mod tests;
...
+		type Origin: From<<Self as SystemConfig>::Origin>
+			+ Into<Result<CumulusOrigin, <Self as Config>::Origin>>;
+
+		type Call: From<Call<Self>> + Encode;
+
+		type XcmSender: SendXcm;
+		type XcmExecutor: ExecuteXcm<<Self as pallet::Config>::Call>;
+		// type WeightInfo: pallet_automation_time::WeightInfo;
+		
+		type Weigher: WeightBounds<<Self as pallet::Config>::Call>;
```
Comment (rhuttman) [inline pallets/template/src/lib.rs:53]: I don't see this being used anywhere

File: pallets/template/src/lib.rs [hunk 9]
```diff
@@ -14,92 +22,155 @@ mod tests;
...
+		
+		type AccountIdToMultiLocation: Convert<Self::AccountId, MultiLocation>;
```
Comment (rhuttman) [inline pallets/template/src/lib.rs:55]: I don't see this being used anywhere

File: pallets/template/src/lib.rs [hunk 10]
```diff
@@ -14,92 +22,155 @@ mod tests;
...
+		type AccountIdToU8Vec: Convert<Self::AccountId, [u8; 32]>;
```
Comment (rhuttman) [inline pallets/template/src/lib.rs:56]: I don't see this being used anywhere

**Review by rhuttman — CHANGES_REQUESTED** [review]

File: pallets/template/src/xcm_test.rs (hunk 1 above)
Comment (justinzhou93) [reply pallets/template/src/xcm_test.rs:3]: nope, just needed for the remark event extrinsic, which i just tossed

File: pallets/template/src/lib.rs (hunk 3 above)
Comment (justinzhou93) [reply pallets/template/src/lib.rs:86]: removed

File: pallets/template/src/lib.rs (hunk 10 above)
Comment (justinzhou93) [reply pallets/template/src/lib.rs:56]: needed here: https://github.com/OAK-Foundation/substrate-parachain-template/pull/3/files#diff-0ec06ea58bd455f09ce6b3bb4c2c1c0d37bda51c1e1be2151c560c9c973959ecR487

File: pallets/template/src/lib.rs (hunk 7 above)
Comment (justinzhou93) [reply pallets/template/src/lib.rs:157]: yessir

**Comment by justinzhou93** [comment]
image
image
image

File: pallets/template/src/lib.rs (hunk 10 above)
Comment (rhuttman) [reply pallets/template/src/lib.rs:49]: it's used by the OakXcmInstructionGenerator but not in this file. Does it need to be passed in?

File: pallets/template/src/lib.rs [hunk 11]
```diff
@@ -16,90 +24,122 @@ mod benchmarking;
...
-			// Update storage.
-			<Something<T>>::put(something);
-
-			// Emit an event.
-			Self::deposit_event(Event::SomethingStored(something, who));
-			// Return a successful DispatchResultWithPostInfo
-			Ok(().into())
+		pub fn delayed_transfer(origin: OriginFor<T>, source: T::AccountId, dest: T::AccountId, value: BalanceOf<T>) -> DispatchResult {
+			let origin_para_id: ParaId = ensure_sibling_para(<T as Config>::Origin::from(origin))?;
+			info!("Force sending balance, source: {:?}, dest: {:?}, value: {:?}, origin_para_id: {:?}", source, dest, value, origin_para_id);
```
Comment (rhuttman) [inline pallets/template/src/lib.rs:89]: update comment to not say force

File: pallets/template/src/lib.rs [hunk 12]
```diff
@@ -16,90 +24,122 @@ mod benchmarking;
...
+		pub fn send_schedule_xcmp_with_crate(
+			origin: OriginFor<T>,
+			provided_id: Vec<u8>,
+			execution_times: Vec<u64>,
+			dest: T::AccountId,
+			value: BalanceOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let tur_para_id: ParaId = ParaId::from(TURING_PARA_ID);
+			let self_para_id: ParaId = ParaId::from(2001);
```
Comment (rhuttman) [inline pallets/template/src/lib.rs:117]: should be able to get this value from the runtime

File: pallets/template/src/lib.rs [hunk 13]
```diff
@@ -16,90 +24,122 @@ mod benchmarking;
...
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+		pub fn send_schedule_xcmp_with_crate(
+			origin: OriginFor<T>,
+			provided_id: Vec<u8>,
+			execution_times: Vec<u64>,
+			dest: T::AccountId,
+			value: BalanceOf<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let tur_para_id: ParaId = ParaId::from(TURING_PARA_ID);
```
Comment (rhuttman) [inline pallets/template/src/lib.rs:116]: since this value will need to change based on the relay chain we probably want to pass it in as a config. That being said we probably need to do something similar for the OakXcmInstructionGenerator in a later PR

**Review by rhuttman — CHANGES_REQUESTED** [review]

File: runtime/src/lib.rs [hunk 14]
```diff
@@ -14,10 +14,11 @@ use sp_api::impl_runtime_apis;
 use sp_core::{crypto::KeyTypeId, OpaqueMetadata};
 use sp_runtime::{
 	create_runtime_str, generic, impl_opaque_keys,
-	traits::{AccountIdLookup, BlakeTwo256, Block as BlockT, IdentifyAccount, Verify},
+	traits::{AccountIdLookup, BlakeTwo256, Block as BlockT, IdentifyAccount, Verify, Convert},
 	transaction_validity::{TransactionSource, TransactionValidity},
 	ApplyExtrinsicResult, MultiSignature,
 };
+use oak_xcm;
```
[the remark this answers is not in the document]
Comment (justinzhou93) [reply runtime/src/lib.rs:21]: yah to instantiate OakXcmInstructionGenerator to pass into the pallet

File: pallets/template/src/lib.rs (hunk 13 above)
Comment (justinzhou93) [reply pallets/template/src/lib.rs:116]: hmm lemme do this in a separate PR cause then i'll have to update the crate. i wanna get this in so that you can play with your code snippets

**Review by rhuttman — APPROVED** [review]
